### PR TITLE
fix(mmr-api): timing-safe admin key comparison

### LINKED ISSUE #307: Make the MMR API admin-key comparison timing-safe
# Plan 013: Make the MMR API admin-key comparison timing-safe

> **Executor instructions**: Follow this plan step by step. Run every
> verification command and confirm the expected result before moving to the
> next step. If anything in the "STOP conditions" section occurs, stop and
> report — do not improvise. When done, update the status row for this plan
> in `plans/README.md` — unless a reviewer dispatched you and told you they
> maintain the index.
>
> **Drift check (run first)**: `git diff --stat 8d0a85a..HEAD -- mmr-api/middleware/`
> If the file changed since this plan was written, compare the "Current
> state" excerpt against the live code before proceeding; on a mismatch,
> treat it as a STOP condition.

## Status

- **Priority**: P3
- **Effort**: S
- **Risk**: LOW
- **Depends on**: none
- **Category**: security
- **Planned at**: commit `8d0a85a`, 2026-06-12

## Reconciliation — 2026-06-13 (vs main `06919df`, Release #296)

**STILL_VALID — no drift.** `mmr-api/middleware/auth.go` is unchanged in
`4c76be7..06919df` (the only mmr-api changes in the delta were `CHANGELOG.md` and
the release `package.json` version bump). Execute as-is. Drift baseline → `06919df`.

## Why this matters

The Go MMR service authenticates the .NET API via an `X-API-KEY` header
compared with `!=` — a byte-wise short-circuiting comparison whose timing
leaks how many leading bytes of a guess are correct. Remote timing attacks on
in-memory string compares are mostly theoretical over a network, but the fix
is one line (`crypto/subtle.ConstantTimeCompare`), removes the finding class
entirely, and is the kind of hardening reviewers expect on an auth gate. The
current code also re-reads the env var on every request and would behave
fail-open-ish only in one scenario worth pinning with a test: an **empty**
`ADMIN_SECRET` must never authenticate anything.

## Current state

`mmr-api/middleware/auth.go` — the complete file:

```go
package middleware

import (
	"github.com/gin-gonic/gin"
	"net/http"
	"os"
)

func RequireAdminAuth(c *gin.Context) {
	apiKey := c.GetHeader("X-API-KEY")

	if apiKey == "" || apiKey != os.Getenv("ADMIN_SECRET") {
		c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"error": "Unauthorized"})
		return
	}

	c.Next()
}
```

Today an empty `ADMIN_SECRET` + non-empty header is rejected (mismatch), and
an empty header is rejected explicitly — i.e. currently fail-closed; the new
code must preserve that. Tests live under `mmr-api/test/` organized by
package (`api`, `custom`, `mmr`, `openskill`); there is no middleware test
yet. Go version: 1.25 (`mmr-api/go.mod`).

## Commands you will need

| Purpose | Command | Expected on success |
|---------|---------|---------------------|
| Build | `cd mmr-api && go build ./...` | exit 0 |
| Tests | `cd mmr-api && go test ./...` | all pass |
| Vet | `cd mmr-api && go vet ./...` | exit 0 |

## Scope

**In scope**:
- `mmr-api/middleware/auth.go`
- `mmr-api/test/middleware/auth_test.go` (create)
- `.changeset/timing-safe-mmr-api-auth.md` (create)

**Out of scope**:
- Reading the secret once at startup / config restructuring
  (`config.LoadEnv`) — behavior-preserving micro-change only.
- The .NET side's API-key header injection.
- Rate limiting.

## Git workflow

- Branch: `advisor/013-timing-safe-mmr-api-auth`
- Conventional commit, e.g. `fix(mmr-api): timing-safe admin key comparison`
- Do NOT push or open a PR unless the operator instructed it.

### Step 1: Replace the comparison

Rewrite the check in `auth.go`:

```go
import (
	"crypto/subtle"
	"net/http"
	"os"

	"github.com/gin-gonic/gin"
)

func RequireAdminAuth(c *gin.Context) {
	apiKey := c.GetHeader("X-API-KEY")
	secret := os.Getenv("ADMIN_SECRET")

	if apiKey == "" || secret == "" ||
		subtle.ConstantTimeCompare([]byte(apiKey), []byte(secret)) != 1 {
		c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"error": "Unauthorized"})
		return
	}

	c.Next()
}
```

The added `secret == ""` guard makes the fail-closed property explicit
(ConstantTimeCompare returns 1 for two empty slices — without the guard, an
unset secret + empty... the empty-header check covers that, but the explicit
guard removes the reasoning chain).

**Verify**: `cd mmr-api && go build ./... && go vet ./...` → exit 0.

### Step 2: Add middleware tests

Create `mmr-api/test/middleware/auth_test.go` following the style of the
existing tests in `mmr-api/test/api/` (gin test context / httptest recorder).
Cases (use `t.Setenv("ADMIN_SECRET", ...)`):

- correct key → next handler runs (200)
- wrong key → 401
- empty header → 401
- `ADMIN_SECRET` unset/empty + empty header → 401
- `ADMIN_SECRET` unset/empty + non-empty header → 401

**Verify**: `cd mmr-api && go test ./...` → all pass, including 5 new cases.

### Step 3: Changeset

`.changeset/timing-safe-mmr-api-auth.md` with `"mmr-api": patch`.

## Test plan

Step 2's five cases are the test plan; the rest of the Go suite must stay
green (`go test ./...` runs everything).

## Done criteria

- [ ] `grep -n 'ConstantTimeCompare' mmr-api/middleware/auth.go` → 1 match
- [ ] `grep -cn '!= os.Getenv' mmr-api/middleware/auth.go` → 0
- [ ] `cd mmr-api && go build ./... && go vet ./... && go test ./...` all exit 0
- [ ] `.changeset/timing-safe-mmr-api-auth.md` exists
- [ ] No files outside the in-scope list are modified (`git status`)
- [ ] `plans/README.md` status row updated

## STOP conditions

- `auth.go` no longer matches the excerpt.
- The existing test packages use a shared harness that a new
  `test/middleware` package can't join without modification — report the
  structure rather than restructuring it.

## Maintenance notes

- If the service ever grows a second secret/header, extract a
  `constantTimeEquals(a, b string) bool` helper then — premature now.
- Reviewer focus: fail-closed on empty secret; both new guards short-circuit
  BEFORE the compare, which is fine (the secret's length is not what the
  timing attack targets — the comparison content is).

### PR DESCRIPTION
**Pull request by bjarkehs**
Closes #307.

## What

Hardens the MMR API admin-auth gate (`RequireAdminAuth`) against timing-based key enumeration.

- Replaces the short-circuiting `apiKey != os.Getenv("ADMIN_SECRET")` byte comparison with `crypto/subtle.ConstantTimeCompare`.
- Adds an explicit empty-`ADMIN_SECRET` guard so the fail-closed property is obvious (behavior is unchanged — an unset secret already rejected every request).

## Why

A byte-wise `!=` compare short-circuits on the first mismatching byte, leaking via timing how many leading bytes of a guess are correct. Remote exploitation is largely theoretical, but the fix is one line and removes the finding class on an auth gate. See #307 for the full rationale.

## Behavior

The accept condition is provably identical to before (`apiKey != "" && apiKey == secret`), now evaluated in constant time. No API or signature changes.

## Tests

Adds the package's first middleware suite, `mmr-api/test/middleware/auth_test.go` (5 cases):

- correct key → 200 (next handler runs)
- wrong key → 401
- empty header → 401
- empty secret + empty header → 401
- empty secret + non-empty header → 401 (the key regression guard)

`go build ./...`, `go vet ./...`, and `go test ./...` all pass.

## Notes

- `crypto/subtle.ConstantTimeCompare` is constant-time only for equal-length inputs (length still leaks via timing) — the standard, accepted trade-off for this primitive; not worth a digest-compare for an admin secret.
- Reading `ADMIN_SECRET` once at startup was intentionally left out of scope.
- Includes a `mmr-api: patch` changeset.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved API key validation security by implementing constant-time comparison to prevent timing-based enumeration attacks.
  * Enhanced security posture by explicitly rejecting invalid or empty authentication configurations with a fail-closed approach.

* **Tests**
  * Added comprehensive test coverage for API key validation including correct keys, wrong keys, and edge cases.

### CODE REVIEW

**Comment by bjarkehs** [comment]
@coderabbitai review

File: mmr-api/test/middleware/auth_test.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
+package middleware_test
```
[the remark this answers is not in the document]
Comment (bjarkehs) [reply mmr-api/test/middleware/auth_test.go:1]: Keeping `package middleware_test` here. The guideline this flags is the descriptive line in `AGENTS.md` ("organized by package (api, custom, mmr, openskill)"), which documents the buckets that existed rather than restricting new ones — there's no test-taxonomy rule in `.coderabbit.yaml`. `package middleware_test` is the idiomatic external test package for the `middleware` package and mirrors `test/api`'s `package api_test`; renaming it to `package api` inside `test/middleware/` would mismatch the directory. I've updated `AGENTS.md` to include `middleware` in that list (b57989a) so the doc reflects reality. Resolving.
